smarts/core/utils/math.py

```diff
@@ -584,3 +585,55 @@ def running_mean(prev_mean: float, prev_step: int, new_val: float) -> Tuple[floa
     new_step = prev_step + 1
     new_mean = prev_mean + (new_val - prev_mean) / new_step
     return new_mean, new_step
+
+
+def _unique_element_combination(first, second):
+    """Generates a combination of set of values result groupings only contain values from unique
+    indices. Only works if len(first_group) <= len(second_group).
+    _unique_perm('ab', '123') -> a1 b2 a1 b3 a2 b1 a2 b3 a3 b1 a3 b2
+    _unique_perm('abc', '1') -> []
+
+    Args:
+        first (Sequence): A sequence of values.
+        second (Sequence): Another sequence of values.
+
+    Yields:
+        Tuple[Tuple[*Any]]: A set of values similar to permutation.
+    """
+    result = [[a] for a in first]
+    sl = len(second)
+    rl = len(first)
+    perms = list(permutations(range(sl), r=rl))
+    for i, p in enumerate(perms):
+        yield tuple(
+            tuple(result[(i * rl + j) % rl] + [second[idx]]) for j, idx in enumerate(p)
+        )
+
+
+def ordered_combinations(first_group, second_group, default=None):
+    """Generate a product that generates that sets the values. If
+    len(first_group) <= len(second_group) the value is padded.
+
+    padded_product('ab', '123') -> a1 b2 a1 b3 a2 b1 a2 b3 a3 b1 a3 b2
+    padded_product('ab', '1', default="k") -> a1 bk ak b1
```

Comment:
This comment might be a bit misleading
- Should it be called `ordered_combinations` instead of `padded_product`?
- From the comment it looks like it returns a linear sequence rather than a sequence of products

Reply:
I have clarified the utility and return values.

It should return groups of pairs, where the pairs use only unique indices from the contributor sequences, within each resulting group.